File: solutions/cpp/flower-field/1/flower_field.cpp
#include "flower_field.h"

using namespace std;

namespace flower_field {

   vector<string> annotate (vector<string> garden){
        vector<vector<int>> deltas{ {0,-1}, {0,1}, {-1,-1}, {-1,0}, {-1,1}, {1,-1}, {1,0}, {1,1}};
        
        int rows = garden.size();
        if (0 == rows)
            return {};
        vector<string> output(rows);
        
        int cols = garden[0].size();
        if (0 == cols)
            return output;
        
        for(int x = 0; x<rows; x++){
            for(int y = 0; y<cols; y++){
                if('*' == garden[x][y]) 
                    output[x].push_back('*');
                else {
                    int count = 0;
                    for(auto& delta : deltas){
                        int new_x = x + delta[0];
                        int new_y = y + delta[1];
                        if(new_x >= 0 && new_x < rows 
                            && new_y >= 0 && new_y < cols
                            && '*' == garden[new_x][new_y])
                                count++;
                    }
                    output[x].push_back(count == 0 ? ' ' : (char)count + '0');
                }
            }
        }
        return output;
    }    

}  // namespace flower_field

File: solutions/dart/flower-field/1/lib/flower_field.dart
class FlowerField {
  List<String> garden;
  
  FlowerField(this.garden);

  List<String> get annotated{
    if(this.garden.isEmpty)
      return <String>[];
    if(garden[0].length == 0)
      return <String>[''];
    var rows = garden.length;
    var cols = garden[0].length;
    var output = List.generate(rows, (r) => List.generate(cols, (c)=>'', growable : false)); 

    for(var y=0; y<rows; y++){
      for(var x=0; x<cols; x++){
        if(garden[y][x] == '*'){
          output[y][x] = '*';
        }
        else {
          int count = 0;
          for(var dx = -1; dx<=1; dx++){
            for(var dy = -1; dy<=1; dy++){
              var newX = x + dx;
              var newY = y + dy;
              if(newY >= 0 && newY < rows 
                  && newX >= 0 && newX < cols
                  && garden[newY][newX]=='*')
                                count++;
            }}
          output[y][x] = count==0 ? ' ' : count.toString(); 
        }
      }  
    }
    return output.map((row)=>row.join()).toList();
  }
}
